src/transformers/trainer.py

```diff
@@ -534,7 +582,21 @@ def train(self, model_path: Optional[str] = None):
                         self._log(logs)
 
                     if self.args.evaluate_during_training and self.global_step % self.args.eval_steps == 0:
-                        self.evaluate()
+                        results = self.evaluate()
+
+                        if self.args.patience > 0:
+                            # Keep track of best loss to determine if we should stop early
+                            eval_loss = results["eval_loss"]
+                            if not patience_best_eval_loss or eval_loss < patience_best_eval_loss:
+                                patience_evals_without_improvement = 0
+                                patience_best_eval_loss = eval_loss
+                            else:
+                                patience_evals_without_improvement += 1
+                                if patience_evals_without_improvement >= self.args.patience:
+                                    patience_should_stop = True
```

Comment:
Maybe name the flag `stop_training`? This way it can be reused by other components introduced later that might need to abort training as well.